packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,260 @@
+};
+
+export const filterActions = (
+  actions: UserAction[],
+  excludedCustomActions: string[]
+): UserAction[] => {
+  const filteredActions = actions.filter((action) => {
+    let shouldExcludeCustomAction = false;
+    const eventType = action.eventData?.type || '';
+    const eventValue = action.eventData?.value || '';
+
+    if (eventType && eventValue) {
+      shouldExcludeCustomAction =
+        excludedCustomActions.includes(eventType) ||
+        excludedCustomActions.includes(eventValue);
+    }
+    return (
+      action.actionType !== UserActionType.CUSTOM || shouldExcludeCustomAction
+    );
+  });
+  return filteredActions;
+};
+
+export const mapUserActions = (rawActions: rawUserAction[]): UserAction[] => {
+  const mappedUserActions = rawActions.map((rawAction) => {
+    const actionData = JSON.parse(rawAction.value);
+    return {
+      actionType: rawAction.name as UserActionType,
+      timestamp: rawAction.time,
+      eventData: {
+        type: actionData.event_type,
+        value: actionData.event_value,
+      },
+      cause: actionData.cause,
+      searchHub: actionData.origin_level_1,
+      document: {
+        title: actionData.title,
+        uriHash: actionData.uri_hash,
+        // eslint-disable-next-line @cspell/spellchecker
+        contentIdKey: actionData.c_contentidkey,
+        // eslint-disable-next-line @cspell/spellchecker
+        contentIdValue: actionData.c_contentidvalue,
+      },
+      query: actionData.query_expression,
+    };
+  });
+  return mappedUserActions;
+};
+
+export const splitActionsIntoTimelineSessions = (
+  actions: UserAction[],
+  ticketCreationDate: string
+): UserActionTimeline => {
+  const returnTimeline: UserActionTimeline = {
+    precedingSessions: [],
+    session: undefined,
+    followingSessions: [],
+    caseSessionFound: false,
+  };
+
+  let currentSession: UserSession = {
+    start: actions[0].timestamp,
+    end: actions[0].timestamp,
+    actions: [],
+  };
+
+  actions.forEach((action) => {
+    // If the action is part of the same session, we add it to the current session
+    if (isPartOfTheSameSession(action, currentSession.end)) {
+      currentSession.actions.push(action);
+      currentSession.start = action.timestamp;
+    } else {
+      // If the action is not part of the same session, we close the current session
+      if (currentSessionHasTicketCreation(currentSession, ticketCreationDate)) {
+        returnTimeline.session = currentSession;
+        returnTimeline.caseSessionFound = true;
+      } else if (
+        currentSessionIsBeforeCaseCreation(currentSession, ticketCreationDate)
+      ) {
+        returnTimeline.precedingSessions.push(currentSession);
+      } else {
+        returnTimeline.followingSessions.push(currentSession);
+      }
+      currentSession = {
+        start: action.timestamp,
+        end: action.timestamp,
+        actions: [],
+      };
+      currentSession.actions.push(action);
+    }
+  });
```

Comment:
Too complex/hard to read, try to simplify:
- use a for..of loop and break/continue statements
- split into functions with descriptive name.

Reply:
Simplified the logic of that function let me know what you think